superset-frontend/src/SqlLab/components/ResultSet/index.tsx

```diff
@@ -664,180 +599,177 @@ export default class ResultSet extends React.PureComponent<
-          </Button>
-        );
-      }
+    if (tempData && tempData.length === 0) {
+      return <Alert type="warning" message={t('The query returned no data')} />;
     }
-    let trackingUrl;
-    let progressBar;
-    if (query.progress > 0) {
-      progressBar = (
-        <ProgressBar
-          percent={parseInt(query.progress.toFixed(0), 10)}
-          striped
-        />
+  }
+  if (query.cached || (query.state === 'success' && !query.results)) {
+    if (query.isDataPreview) {
+      return (
+        <Button
+          buttonSize="small"
+          buttonStyle="primary"
+          onClick={() =>
+            reFetchQueryResults({
+              ...query,
+              isDataPreview: true,
+            })
+          }
+        >
+          {t('Fetch data preview')}
+        </Button>
       );
     }
-    if (query.trackingUrl) {
-      trackingUrl = (
+    if (query.resultsKey) {
+      return (
         <Button
           buttonSize="small"
-          onClick={() => query.trackingUrl && window.open(query.trackingUrl)}
+          buttonStyle="primary"
+          onClick={() => fetchResults(query)}
         >
-          {t('Track job')}
+          {t('Refetch results')}
         </Button>
       );
     }
-    const progressMsg =
-      query && query.extra && query.extra.progress
-        ? query.extra.progress
-        : null;
-
-    return (
-      <div style={LOADING_STYLES}>
-        <div>{!progressBar && <Loading position="normal" />}</div>
-        {/* show loading bar whenever progress bar is completed but needs time to render */}
-        <div>{query.progress === 100 && <Loading position="normal" />}</div>
-        <QueryStateLabel query={query} />
-        <div>
-          {progressMsg && <Alert type="success" message={progressMsg} />}
-        </div>
-        <div>{query.progress !== 100 && progressBar}</div>
-        <div>{trackingUrl}</div>
-      </div>
+  }
+  let trackingUrl;
+  let progressBar;
+  if (query.progress > 0) {
+    progressBar = (
+      <ProgressBar percent={parseInt(query.progress.toFixed(0), 10)} striped />
     );
   }
-}
+  if (query.trackingUrl) {
+    trackingUrl = (
+      <Button
+        buttonSize="small"
+        onClick={() => query.trackingUrl && window.open(query.trackingUrl)}
+      >
+        {t('Track job')}
+      </Button>
+    );
+  }
+  const progressMsg =
+    query && query.extra && query.extra.progress ? query.extra.progress : null;
+
+  return (
+    <div style={LOADING_STYLES}>
+      <div>{!progressBar && <Loading position="normal" />}</div>
+      {/* show loading bar whenever progress bar is completed but needs time to render */}
+      <div>{query.progress === 100 && <Loading position="normal" />}</div>
+      <QueryStateLabel query={query} />
+      <div>{progressMsg && <Alert type="success" message={progressMsg} />}</div>
+      <div>{query.progress !== 100 && progressBar}</div>
+      <div>{trackingUrl}</div>
+    </div>
+  );
+};
+
+ResultSet.defaultProps = {
```

Comment:
I think you can move this in favor of stating defaults in the functional component itself.